openactive/openactive-test-suite
# Feed Snapshot Checker: Orchestration Script

Changes:

* Created an orchestration script: `src/scripts/doFeedSnapshotValidation.js`. This:
    * Saves feed snapshot (using broker)
    * Checks if there is a previous snapshot to compare to
    * If there is, does the comparison
* Feed Snapshot Validator: Added a comparison function: `checkModificationsArePushedToTheEndOfTheFeed(..)`, which I believe covers all of these checks we had discussed before:
    > * A new item that wasn't in the feed before is added to a place earlier than the previous "last" position in the feed
    > * An existing item that wasn't deleted disappears from the feed (unless it is allowed to be deleted as a result of being very old - there's some logic for this somewhere in the docs)
    > * An existing item isn't changed while staying in the same place

    Also, added single-feed comparison function `checkRpdeOrder(..)` which asserts that all items are ordered correctly. This is a pre-condition to the `checkModificationsArePushedToTheEndOfTheFeed(..)` being able to verify correctly
* Feed Snapshot Validator: This now takes "earlier" and "later" file paths as inputs, rather than defining them itself. This hopefully makes it easier to flexibly compare between different snapshots
* Broker: Add flag which makes Broker stop once its finished its initial harvest `--exit-at-end-of-feeds`
* Broker: Saves snapshots with a timestamp rather than `previous` / `latest`. Hopefully this makes it easier to do comparisons across different times
* Broker: Add script to remove old feed snapshots (as the previous change means that you can accumulate snapshots as the "previous" snapshot is no longer deleted)
* Some documentation but definitely not complete

This isn't ready to merge into Test Suite yet unfortunately. There are some unanswered questions and incomplete documentation

## QA

An example run on WePlay Football's staging feed (Used `NO_VALIDATE=true` to speed up the snapshot saving):

```sh
➜  openactive-test-suite git:(feature/feed-snapshot-checker) ✗ NODE_ENV=dev NO_VALIDATE=true node src/scripts/doFeedSnapshotValidation.js

> openactive-broker-microservice@1.0.0 save-feed-snapshot-only
> node app.js --save-feed-snapshot --exit-at-end-of-feeds

Broker Microservice running on port 3000

Check http://localhost:3000/status for current harvesting status

Downloading Dataset Site JSON-LD from "https://partners-staging.weplayfootball.com/OpenActive"...
Dataset Site JSON-LD: {
  "@context": [
    "https://schema.org/",
    "https://openactive.io/"
  ],
  "@type": "Dataset",
  "@id": "https://partners-staging.weplayfootball.com/OpenActive/",
  "name": "PlayFootball Facilities",
  "description": "Near real-time availability and rich descriptions relating to the facilities available from PlayFootball",
  "accessService": {
    "@type": "WebAPI",
    "name": "Open Booking API",
    "description": "API that allows for seamless booking experiences to be created for facilities available from PlayFootball",
    "conformsTo": [
      "https://openactive.io/open-booking-api/EditorsDraft/"
    ],
    "documentation": "https://permalink.openactive.io/dataset-site/open-booking-api-documentation",
    "endpointDescription": "https://www.openactive.io/open-booking-api/EditorsDraft/swagger.json",
    "endpointUrl": "https://partners-staging.weplayfootball.com/OpenActive/api",
    "landingPage": "https://forms.office.com/r/2xF6J6ih4V"
  },
  "backgroundImage": {
    "@type": "ImageObject",
    "url": "https://partners-staging.weplayfootball.com/images/placeholder-dataset-site-background.jpg"
  },
  "bookingService": {
    "@type": "BookingService",
    "name": "Spawtz",
    "hasCredential": "https://www.spawtz.com/openactive/",
    "url": "https://www.spawtz.com/"
  },
  "dateModified": "2022-10-20T18:35:53+01:00",
  "datePublished": "2019-01-14T00:00:00+00:00",
  "discussionUrl": "https://github.com/openactive/OpenActive.Server.NET/issues",
  "distribution": [
    {
      "@type": "DataDownload",
      "identifier": "FacilityUse",
      "name": "FacilityUse",
      "additionalType": "https://openactive.io/FacilityUse",
      "contentUrl": "https://partners-staging.weplayfootball.com/OpenActive/feeds/facility-uses",
      "encodingFormat": "application/vnd.openactive.rpde+json; version=1"
    },
    {
      "@type": "DataDownload",
      "identifier": "FacilityUseSlot",
      "name": "Slot for FacilityUse",
      "additionalType": "https://openactive.io/Slot",
      "contentUrl": "https://partners-staging.weplayfootball.com/OpenActive/feeds/facility-use-slots",
      "encodingFormat": "application/vnd.openactive.rpde+json; version=1"
    }
  ],
  "documentation": "https://permalink.openactive.io/dataset-site/open-data-documentation",
  "inLanguage": [
    "en-GB"
  ],
  "keywords": [
    "Facilities",
    "Activities",
    "Sports",
    "Physical Activity",
    "OpenActive"
  ],
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "publisher": {
    "@type": "Organization",
    "name": "PlayFootball",
    "email": "no-reply@better.weplayfootball.com",
    "legalName": "PlayFootball",
    "logo": {
      "@type": "ImageObject",
      "url": "https://weplayfootball.com/images/weplayfootball.svg"
    },
    "url": "https://partners-staging.weplayfootball.com"
  },
  "schemaVersion": "https://openactive.io/modelling-opportunity-data/2.0/",
  "url": "https://partners-staging.weplayfootball.com/OpenActive/"
}

Warning: Open ID Connect Identity Server (accessService.authenticationAuthority) not found in dataset site
Found parent opportunity feed: https://partners-staging.weplayfootball.com/OpenActive/feeds/facility-uses
Found opportunity feed: https://partners-staging.weplayfootball.com/OpenActive/feeds/facility-use-slots

Blocking integration tests to wait for harvest completion...
FacilityUse | 1 items harvested from 1 pages | Response time: 43ms | Elapsed: 1s | Validated: 0 of 1 (0%) ETA: NFs | Status: Harvesting Complete, Validating...
FacilityUseSlot | 720 items harvested from 2 pages | Response time: 51ms | Elapsed: 1s | Validated: 0 of 645 (0%) ETA: NFs | Status: Harvesting Complete, Validating...
Harvesting is up-to-date

> openactive-feed-snapshot-validator@1.0.0 start
> node --unhandled-rejections=strict src/app.js "/Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_173220" "/Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_183552"

SINGLE-FEED PASS: /Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_173220/FacilityUse.json
SINGLE-FEED PASS: /Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_183552/FacilityUse.json
SINGLE-FEED PASS: /Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_173220/FacilityUseSlot.json
SINGLE-FEED PASS: /Users/lukewinship/Dev/openactive/openactive-test-suite/packages/openactive-broker-microservice/feed_snapshots/https%3A%2F%2Fpartners-staging.weplayfootball.com%2FOpenActive/20221020_183552/FacilityUseSlot.json
COMPARISON PASS: FacilityUse.json
COMPARISON PASS: FacilityUseSlot.json
➜  openactive-test-suite git:(feature/feed-snapshot-checker)
```
